### PR TITLE
fix(ext/http): Ensure Deno.serve works across --watch restarts

### LINKED ISSUE #16699: Uncaught AddrInUse when restarting server in watch mode after saving file
Source: https://stackoverflow.com/questions/74444443/deno-watcher-uncaught-addrinuse-only-one-usage-of-each-socket-address

Reproduction:

- Run the code below with `deno run -A --watch server.ts`:

```ts
import express, { Express, Request, Response } from "npm:express@4.18";
import { v4 as uuidv4 } from 'npm:uuid@9.0.0';
const app: Express = express();

app.use("/", (_req: Request, res: Response) => {
    res.send(`uuid: ${uuidv4()}.`);
})

app.listen(8000, () => {
    console.log("App Running: http://localhost:8000");
});
```

- Make a change and save the file (either with auto-save or manually)
- When the file watcher restarts the error is thrown

```
Watcher File change detected! Restarting!
error: Uncaught AddrInUse: Only one usage of each socket address (protocol/network address/port) is normally permitted. (os error 10048)
Watcher Process finished. Restarting on file change...
```

I found that this error can happen when there is another process listening on the same port: https://stackoverflow.com/a/62190276/12769288, so maybe the file watcher tries to restart the server while the previous one is still listening?

Additionally the same error seems to happen on occasion in one of the tests: https://github.com/denoland/deno/issues/16044

### PR DESCRIPTION
**Pull request by mmastrac**
Fixes #16699 and #18960 by ensuring that we release our HTTP `spawn_local` tasks when the HTTP resource is dropped.

Because our cancel handle was being projected from the resource via `RcMap`, the resource was never `Drop`ped. By splitting the handle out into its own `Rc`, we can avoid keeping the resource alive and let it drop to cancel everything.

### CODE REVIEW

**Comment by bartlomieju** [comment]
PR LGTM, but I belive we should add a test in `watcher_tests.rs` to ensure we never hit this problem again.

**Review by aapoalas — DISMISSED** [review]
Added test fails :(

File: cli/util/file_watcher.rs [hunk 1]
```diff
@@ -304,6 +304,9 @@ where
   }
 
   loop {
+    // We may need to give the runtime a tick to settle, as cancellations may need to propagate to tasks
+    tokio::time::sleep(Duration::from_millis(1)).await;
```
Comment (mmastrac) [inline cli/util/file_watcher.rs:308]: I'm not really happy with this part of the PR -- but it's very difficult to get at the resource table from the watcher. Ideally we should be shutting down all resources in the resource table _if_ we're in watch mode, and letting things drop if we're in a normal run mode.

Comment (bartlomieju) [reply cli/util/file_watcher.rs:308]: What is the problem with dropping `OpState` and `ResourceTable` in the watcher - if it's done properly I'd expect everything to be closed done cleanly. Then we can `yield()` 10 times after the drop to make sure that Tokio has enough ticks to resolve everything.

Comment (mmastrac) [reply cli/util/file_watcher.rs:308]: I think to be really "correct" we should be explicitly `close`ing all the resources in the resource table on watch restart to ensure that all cancellations are triggered and propagate correctly to futures, including those inside of tasks. As it stands right now, we don't really have a good way to 1) start the close process for every resource and 2) confirm that the close process for every resource has completed.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Comment by thaysang** [comment]
same problem when -> import express from 'npm:express'
deno 1.33.2 (release, x86_64-pc-windows-msvc)
v8 11.4.183.1
typescript 5.0.3

**Pull request merged**